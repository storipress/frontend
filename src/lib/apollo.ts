import type {
  ApolloClient,
  ApolloError,
  ApolloQueryResult,
  DocumentNode,
  FetchResult,
  NormalizedCacheObject,
  OperationVariables,
  SubscribeToMoreOptions,
  TypedDocumentNode,
  WatchQueryOptions,
} from '@apollo/client/core'
import { noop } from 'lodash-es'
import type {
  UseQueryReturn as ApolloUseQueryrReturn,
  MutateOverrideOptions,
  UseMutationOptions,
} from '@vue/apollo-composable'
import {
  provideApolloClient,
  useLazyQuery as useApolloLazyQuery,
  useMutation as useApolloMutation,
  useSubscription as useApolloSubscription,
} from '@vue/apollo-composable'
import { createEventHook } from '@vueuse/core'
import type { Ref } from 'vue'
import { isRef, ref, unref } from 'vue'
import type { MaybeRef, MaybeRefOrGetter } from '@vueuse/core'
import { resolveFeatureFlag } from './feature-flag'
import { useApolloClientID } from './client-id'
import { CANCEL_ERROR } from '~/api/blocking-link'
import { createApolloClient } from '~/api/client'
import { useAuthStore } from '~/stores/auth'
import { env } from '~/env'

const API_HOST = env.VITE_ENABLE_MOCK_API ? 'http://localhost:4000' : env.VITE_API_HOST
const DEFAULT_URL = `${API_HOST}/graphql`

export const clients: Record<string, ApolloClient<NormalizedCacheObject>> = {
  default: createApolloClient(DEFAULT_URL),
}
clients._ = clients.default

interface UseApolloClientReturn {
  client: ApolloClient<NormalizedCacheObject>
  clientID: string
}

export function getApolloClient(clientID: string) {
  clients[clientID] ??= createApolloClient(`${env.VITE_API_HOST}/client/${clientID}/graphql`, clientID)
  return clients[clientID] || clients.default
}

export function useApolloClient(): UseApolloClientReturn {
  const store = useAuthStore()
  const clientID = useApolloClientID()
  store.clientID = clientID
  const client = getApolloClient(clientID)
  return { client, clientID }
}

export function query<TResult, TVariables = undefined>(
  clientID: string,
  query: DocumentNode | TypedDocumentNode<TResult, TVariables>,
  variables?: TVariables,
  options = {},
): Promise<ApolloQueryResult<TResult>> {
  clients[clientID] ??= createApolloClient(`${env.VITE_API_HOST}/client/${clientID}/graphql`, clientID)
  const client = clients[clientID] || clients.default
  return client.query({ ...options, query, variables })
}

export function mutate<TResult, TVariables extends OperationVariables | undefined = undefined>(
  clientID: string,
  mutation: DocumentNode | TypedDocumentNode<TResult, TVariables>,
  variables?: TVariables,
  options = {},
): Promise<FetchResult<TResult>> {
  clients[clientID] ??= createApolloClient(`${env.VITE_API_HOST}/client/${clientID}/graphql`, clientID)
  const client = clients[clientID] || clients.default
  return client.mutate({ ...options, mutation, variables })
}

type ToReactiveSource<T> = MaybeRef<T> | (() => T)

// cache-and-network fetchPolicy is not working with useQuery in @vue/apollo-composable
export function useQuery<TResult, TVariables = undefined>(
  query: DocumentNode | TypedDocumentNode<TResult, TVariables>,
  variables?: MaybeRef<TVariables>,
  options: Omit<WatchQueryOptions<TVariables, TResult>, 'query' | 'variables'> = {},
): Pick<ApolloUseQueryrReturn<TResult, TVariables>, 'result' | 'refetch' | 'loading' | 'subscribeToMore'> {
  const { client } = useApolloClient()
  // must use ref for deep watch
  const result = ref<TResult>()
  const loading = ref<boolean>(false)
  const variablesRef: Ref<TVariables> = isRef<TVariables>(variables)
    ? variables
    : (ref<TVariables>(variables as TVariables) as Ref<TVariables>)

  const observableQuery = client.watchQuery({ ...options, query, variables: unref(variables) })
  loading.value = true
  observableQuery.subscribe(({ data }) => {
    result.value = data
    loading.value = false
    // data maybe same reference because of Apollo cache
    triggerRef(result)
  })

  watch(variablesRef, async (variables) => {
    loading.value = true
    await observableQuery.refetch(variables)
    loading.value = false
  })

  return {
    result,
    loading,
    refetch: async (variables?: TVariables) => {
      loading.value = true
      const queryResult = await observableQuery.refetch(variables)
      loading.value = false
      return queryResult
    },
    subscribeToMore<TSubscribeVariables = TVariables, TSubscribeResult = TResult>(
      options: ToReactiveSource<SubscribeToMoreOptions<TResult, TSubscribeVariables, TSubscribeResult>>,
    ) {
      let stop = noop
      const reactiveOptions = typeof options === 'function' ? options : ref(options)
      watch(
        reactiveOptions,
        (options, _old, onCleanup) => {
          stop = observableQuery.subscribeToMore(
            options as SubscribeToMoreOptions<TResult, TSubscribeVariables, TSubscribeResult>,
          )
          onCleanup(stop)
        },
        { immediate: true },
      )

      return () => {
        stop()
      }
    },
  }
}

export const useLazyQuery: typeof useApolloLazyQuery = <TResult, TVariables extends OperationVariables>(
  query: MaybeRef<DocumentNode | TypedDocumentNode<TResult, TVariables>>,
  variables?: MaybeRefOrGetter<TVariables>,
  options: MaybeRefOrGetter<Omit<WatchQueryOptions<TVariables, TResult>, 'query' | 'variables'>> = {},
) => {
  const { clientID, client } = useApolloClient()
  provideApolloClient(client)
  return useApolloLazyQuery(toValue(query), variables, { ...options, clientId: clientID })
}

/**
 * Conditionally load a query
 * ! Notice: It's useless to test it with Growthbook Devtool, because override value won't apply on the first request
 * @param featureFlag Feature flag key or feature flag
 * @param query Query
 * @param variables Query variables
 * @param options Query options
 */
export function useFeatureFlaggedQuery<TResult, TVariables = undefined>(
  featureFlag: MaybeRefOrGetter<boolean> | string,
  query: MaybeRefOrGetter<DocumentNode | TypedDocumentNode<TResult, TVariables>>,
  variables?: MaybeRefOrGetter<TVariables>,
  options: MaybeRefOrGetter<Omit<WatchQueryOptions<TVariables, TResult>, 'query' | 'variables'>> = {},
) {
  const { clientID, client } = useApolloClient()
  provideApolloClient(client)
  const res = useApolloLazyQuery(query, variables, { ...options, clientId: clientID })
  const flag = resolveFeatureFlag(featureFlag)
  whenever(
    flag,
    () => {
      res.load()
    },
    { immediate: true },
  )
  return {
    ...res,
    load() {
      if (!flag.value) {
        return
      }
      return res.load()
    },
    refetch() {
      if (!flag.value) {
        return
      }
      return res.refetch()
    },
  }
}

export const useMutation: typeof useApolloMutation = (document: any, options = {}) => {
  const { clientID, client } = useApolloClient()
  if (env.MODE !== 'test') {
    provideApolloClient(client)
  }
  return useApolloMutation(document, { ...options, clientId: clientID })
}

/**
 * This is similar to useMutation, but it accept a extra Promise as the first argument to indicate the cancellation.
 * @param document GraphQL mutation document
 * @param options Apollo mutation options
 * @returns
 */
export function useCancelableMutation<TResult = any, TVariables extends OperationVariables = OperationVariables>(
  document: TypedDocumentNode<TResult, TVariables>,
  options: UseMutationOptions<TResult, TVariables> = {},
) {
  const { clientID, client } = useApolloClient()
  provideApolloClient(client)
  const cancelEvent = createEventHook<void>()
  const res = useApolloMutation(document, { ...options, clientId: clientID })
  return {
    ...res,
    onCancel: cancelEvent.on,
    mutate: (
      cancelPromise: Promise<boolean>,
      variables?: TVariables | null | undefined,
      overrideOptions: MutateOverrideOptions<TResult> = {},
    ) => {
      return res
        .mutate(variables, {
          ...overrideOptions,
          context: {
            ...overrideOptions.context,
            blocking: cancelPromise,
          },
        })
        .catch((err: ApolloError) => {
          // Swallow cancellation errors
          if (err.networkError && err.networkError === CANCEL_ERROR) {
            cancelEvent.trigger()
            return
          }
          throw err
        })
    },
  }
}

export const useSubscription: typeof useApolloSubscription = ((document: any, variables?: any, options = {}) => {
  const { clientID, client } = useApolloClient()
  provideApolloClient(client)
  return useApolloSubscription(document, variables, {
    ...options,
    clientId: clientID,
  })
}) as any
