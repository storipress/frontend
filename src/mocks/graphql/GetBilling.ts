import { defineGraphQLMock } from '../define-graphql-mock'
import { GetBillingDocument } from '~/graphql-operations'

export default defineGraphQLMock(GetBillingDocument, {
  billing: {
    id: '1',
    on_trial: false,
    canceled: false,
    credit_balance: '0',
    has_historical_subscriptions: false,
    has_pm: true,
    next_pm_discounts: [],
    next_pm_taxes: [],
    on_grace_period: false,
    publications_count: 1,
    publications_quota: 99,
    seats_in_use: 1,
    subscribed: true,
    next_pm_date: '2020-01-01',
    trial_ends_at: null,
    ends_at: null,
    source: 'stripe',
    interval: 'monthly',
    next_pm_subtotal: '0',
    next_pm_tax: '0',
    next_pm_total: '0',
    plan: 'publisher',
    plan_id: '1',
    pm_last_four: '0000',
    pm_type: 'visa',
    quantity: 1,
    referer: null,
    has_prophet: false,
  },
})
