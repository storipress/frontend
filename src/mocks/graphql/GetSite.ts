import { defineGraphQLMock } from '../define-graphql-mock'
import { GetSiteDocument, SubscriptionSetup } from '~/graphql-operations'

export default defineGraphQLMock(GetSiteDocument, {
  site: {
    id: 'D6RX98VXN',
    name: "Jack's Blog",
    description: 'work hard, play hard',
    favicon: null,
    lang: 'en-US',
    logo: {
      token: '',
      url: '',
      name: '',
      mime: '',
      size: 0,
      width: 0,
      height: 0,
      title: '',
      caption: '',
      description: '',
      transformation: '',
    },
    workspace: 'joesblog',
    socials: '{"Facebook":"www.facebook.com\\/Storipress","Instagram":"test"}',
    timezone: 'Pacific/Honolulu',
    custom_domain: 'www.joesblog.storipress.dev',
    customer_site_domain: 'www.joesblog.storipress.dev',
    customer_site_storipress_url: 'joesblog-cdn.storipress.dev',
    site_domain: 'www.joesblog.storipress.dev',
    mail_domain: 'joesblog.storipress.dev',
    plan: 'publisher',
    subscription: false,
    newsletter: true,
    email: 'test@storipress.com',
    currency: 'AUD',
    monthly_price: '20',
    yearly_price: '100',
    custom_domain_email: [
      {
        hostname: '20220422081255pm._domainkey.www.ksjdfhksdhfkasjhdfhsjakf.com',
        type: 'TXT',
        value:
          'k=rsa; p=MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCJxCP6xGRXSZEMVQAx4kr87A6kYM6nwY3RTLUKS9Ne028RbcDmbvNrey9P7mm6Ma0hdh1zbKifLIyt5p3jvfX2yeCuJ8+nLNnzz0G6viEvMflhKCcbBycn6qZOzLFakt+RKtQRSeyr0cWzH/PUwYgPOEZVRuk04AGdI8gIDWvRLwIDAQAB',
        __typename: 'EmailDNSRecord',
      },
      {
        hostname: 'pm-bounces.www.ksjdfhksdhfkasjhdfhsjakf.com',
        type: 'CNAME',
        value: 'pm.mtasv.net',
        __typename: 'EmailDNSRecord',
      },
    ],
    subscription_setup: SubscriptionSetup.Done,
    __typename: 'Site',
    has_prophet: false,
    prophet_config: null,
    paywall_config: null,
  },
})
