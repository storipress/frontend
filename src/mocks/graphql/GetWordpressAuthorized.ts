import { defineGraphQLMock } from '../define-graphql-mock'
import { GetWordpressAuthorizedDocument } from '~/graphql-operations'

export default defineGraphQLMock(GetWordpressAuthorizedDocument, {
  wordPressAuthorized: true,
})
