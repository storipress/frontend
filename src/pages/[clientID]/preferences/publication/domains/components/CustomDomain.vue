<script setup lang="ts">
import { Alert, Buttons, Icon, Inputs, NOTIFICATION_KEY } from '@storipress/core-component'
import * as Sentry from '@sentry/vue'
import {
  DNS_SETTINGS_MISCONFIGURED,
  DNS_SETTINGS_SUCCESSFUL,
  EMAIL_RECORDS_INCORRECT,
  INCORRECT_HOST_NAME,
} from '../notifications'
import CustomDomainRecordsForm from './CustomDomainRecordsForm.vue'
import LoadingRecords from './LoadingRecords.vue'
import DisconnectModal from './DisconnectModal.vue'
import { Flags, useFeatureFlag } from '~/lib/feature-flag'
import { errorResponse, useCustomDomain, useTutorials, useUserSubscription } from '~/composables'
import type { GetSiteQuery } from '~/graphql-operations'
import { useSiteStore } from '~/stores/site'

const props = defineProps<{
  site: GetSiteQuery['site']
}>()

const {
  customDomain,
  customDomainAvailability,
  customDomainDnsStatus,
  isInvalidDomain,
  status,
  loading,
  checkCustomDomainAvailability,
  getDomainConfiguration,
  checkCustomDomainDnsStatus,
  checkAllCustomDomainDnsStatus,
  checkIncorrectHostWithDNSQuery,
  activateCustomDomain,
  removeCustomDomain,
  parseDomainResult,
} = useCustomDomain()

const enabledPaidCustomDomain = useFeatureFlag(Flags.PaidCustomDomain)
const { onTrialFree } = useUserSubscription()
const plan = ref('free')
const isFreePlan = computed(() => plan.value === 'free')

const notifications = inject(NOTIFICATION_KEY)
const siteStore = useSiteStore()
const { setTutorials } = useTutorials()

checkCustomDomainDnsStatus()

const canGetDomainConfiguration = computed(() => {
  return customDomainAvailability?.mail && customDomainAvailability?.site
})
const domainName = computed({
  get() {
    return customDomain.site
  },
  set(domainName) {
    const result = parseDomainResult(domainName)
    if (result?.isSubDomain && !domainName.startsWith('www.')) {
      Object.assign(customDomain, {
        site: domainName,
        mail: domainName,
        redirect: [],
      })
    } else {
      const { isNakedDomain, nakedDomain, nonNakedDomain } = checkNakedDomain(domainName)

      Object.assign(customDomain, {
        site: domainName,
        mail: nakedDomain,
        redirect: [isNakedDomain ? nonNakedDomain : nakedDomain],
      })
    }
  },
})
const domainRecords = computed(() => {
  const mergeRecords = [
    ...(customDomainDnsStatus.site ?? []),
    ...(customDomainDnsStatus.redirect ?? []),
    ...(customDomainDnsStatus.mail ?? []),
  ]
  return mergeRecords
})
const rootDomain = computed(() => parseDomainResult(customDomainDnsStatus.site[0].domain)?.rootDomain ?? '')

const hasSetupDomain = computed(() => siteStore.site?.mail_domain && siteStore.site?.site_domain)
const hasChanged = computed(() => {
  return Boolean(domainRecords.value.length) && customDomainDnsStatus?.site?.[0]?.domain !== customDomain.site
})
const unavailableForConnect = computed(() => {
  return Boolean(hasSetupDomain.value) || (enabledPaidCustomDomain.value && isFreePlan.value && !onTrialFree.value)
})

const getDomainConfigurationLoading = ref(false)
async function onGetDomainConfiguration() {
  getDomainConfigurationLoading.value = true
  if (hasChanged.value) {
    const removeSuccess = await removeCustomDomain()
    // confirm that has been completely removed before re-initialize
    if (!removeSuccess) {
      Sentry.captureException(new Error('fail to remove custom domain'))
      return
    }
  }
  await getDomainConfiguration()
  getDomainConfigurationLoading.value = false
}

function checkNakedDomain(domain: string) {
  if (domain.startsWith('www.')) {
    return {
      isNakedDomain: false,
      nakedDomain: domain.replace(/^(www\.)/, ''),
    }
  } else {
    return {
      isNakedDomain: true,
      nakedDomain: domain,
      nonNakedDomain: `www.${domain}`,
    }
  }
}

function onInput(e: { target: HTMLInputElement }) {
  loading.value = true
  domainName.value = e.target.value
  checkCustomDomainAvailability(domainName.value)
}
async function onConfirmCustomDomain() {
  if (hasSetupDomain.value) {
    return
  }

  await checkCustomDomainDnsStatus()
  checkAllCustomDomainDnsStatus({ site: customDomainDnsStatus.site, mail: customDomainDnsStatus.mail })
  if (!status.all) {
    const isIncorrectHostName = await checkIncorrectHostWithDNSQuery(customDomainDnsStatus.site[0])

    if (status.error.site === errorResponse.notFound && isIncorrectHostName) {
      const { type, domain } = customDomainDnsStatus.site[0]
      notifications?.({
        title: INCORRECT_HOST_NAME.title.replace('__DOMAIN_TYPE__', type),
        type: 'warning',
        iconName: 'warning',
        content: INCORRECT_HOST_NAME.content.replace('__DOMAIN_TYPE__', type).replace('__REPLACE_DOMAIN__', domain),
      })
      return
    }

    notifications?.({
      title: DNS_SETTINGS_MISCONFIGURED.title,
      type: 'warning',
      iconName: 'warning',
      content: DNS_SETTINGS_MISCONFIGURED.content,
    })
  } else if (status.siteDomain && !status.mailDomain) {
    notifications?.({
      title: EMAIL_RECORDS_INCORRECT.title,
      type: 'warning',
      iconName: 'warning',
      content: EMAIL_RECORDS_INCORRECT.content,
    })
  } else {
    const result = await activateCustomDomain()
    if (result) {
      notifications?.({
        title: DNS_SETTINGS_SUCCESSFUL.title,
        type: 'primary',
        content: DNS_SETTINGS_SUCCESSFUL.content,
      })
      setTutorials('setDomain')
    }
    siteStore.fetchSite()
  }
}

const disconnectModalOpen = ref(false)
const isLoading = ref(false)
async function onDisconnect() {
  isLoading.value = true
  const removeSuccess = await removeCustomDomain()
  if (!removeSuccess) {
    Sentry.captureException(new Error('fail to remove custom domain'))
    return
  }
  await checkCustomDomainDnsStatus()
  await siteStore.fetchSite()
  isLoading.value = false
  disconnectModalOpen.value = false
}

watch(
  () => props.site,
  (data) => {
    plan.value = data?.plan ?? 'free'
  },
  { immediate: true },
)
</script>

<template>
  <Alert
    v-if="enabledPaidCustomDomain && isFreePlan && !onTrialFree"
    message="Upgrade account to connect a custom domain."
    type="warning"
    class="mb-8"
  >
    <template #description>
      <span class="text-body text-stone-800">
        Connect a custom domain, collect revenue on members, invite more Admin Editors, connect to Shopify and Webflow,
        and more.
      </span>
    </template>
  </Alert>

  <form @submit.prevent>
    <div class="flex items-end gap-x-2">
      <Inputs
        v-model="domainName"
        label="1. Enter your domain name"
        placeholder="Enter the domain you want to connect"
        html-name="customDomain"
        html-type="text"
        :disabled="unavailableForConnect"
        :show-error="Boolean(domainName)"
        class="w-full"
        @input="onInput"
      />
      <Buttons
        type="main"
        color="primary"
        html-type="submit"
        class="size-9 transition-colors duration-200 ease-in-out"
        :disabled="!canGetDomainConfiguration || isInvalidDomain || (Boolean(domainRecords.length) && !hasChanged)"
        @click="onGetDomainConfiguration"
      >
        <Icon
          icon-name="check"
          class="scale-[.6] rounded-full p-1.5"
          :class="
            !canGetDomainConfiguration || isInvalidDomain || (Boolean(domainRecords.length) && !hasChanged)
              ? 'bg-stone-600 text-stone-200'
              : 'bg-white text-emerald-700'
          "
        />
      </Buttons>
    </div>

    <LoadingRecords v-if="getDomainConfigurationLoading" class="mt-8" />

    <template v-if="Boolean(domainRecords.length) && !hasChanged">
      <LoadingRecords v-if="loading" class="mt-8" />

      <template v-else>
        <CustomDomainRecordsForm :domain-records="domainRecords" :root-domain="rootDomain" class="my-8" />

        <div class="flex gap-x-2">
          <Buttons
            type="main"
            color="primary"
            html-type="submit"
            :disabled="unavailableForConnect"
            class="h-11 w-full"
            @click="onConfirmCustomDomain"
          >
            {{ hasSetupDomain ? `Custom domain Connected` : `Verify DNS Settings` }}
          </Buttons>
          <Buttons
            v-if="hasSetupDomain"
            type="main"
            color="warning"
            html-type="submit"
            class="size-11 text-base transition-colors duration-200 ease-in-out"
            @click="disconnectModalOpen = true"
          >
            <Icon icon-name="link_break" />
          </Buttons>
        </div>
      </template>
    </template>
  </form>

  <DisconnectModal
    title="Disconnect custom domain"
    info="Are you sure you want to disconnect your custom domain? Your site will be immediately inaccessible via this domain."
    :visible="disconnectModalOpen"
    :loading="isLoading"
    @on-modal-close="disconnectModalOpen = false"
    @on-confirm="onDisconnect"
  />
</template>
