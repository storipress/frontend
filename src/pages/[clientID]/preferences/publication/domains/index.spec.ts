import { beforeEach, describe, expect, it } from 'vitest'
import { fireEvent, waitFor } from '@testing-library/vue'
import { setActivePinia } from 'pinia'
import CustomDomain from './index.vue'
import { enableFeatures, mockResponseOnce, render, setupTestPinia } from '~/test-helpers'
import { Flags } from '~/lib/feature-flag'
import { useApolloClient } from '~/lib/apollo'
import GetBilling from '~/mocks/graphql/GetBilling'
import GetSite from '~/mocks/graphql/GetSite'
import GetWordpressAuthorized from '~/mocks/graphql/GetWordpressAuthorized'

describe('setup publication domain', () => {
  beforeEach(() => {
    enableFeatures([Flags.PaidCustomDomain])
    setActivePinia(setupTestPinia(false))
    const { client } = useApolloClient()
    client.cache.reset()

    // disable wordpress integration
    mockResponseOnce(GetWordpressAuthorized, {
      wordPressAuthorized: false,
    })
  })

  it('render custom domain and email', async () => {
    mockResponseOnce(GetSite, {
      site: {
        plan: 'free',
      },
    })
    mockResponseOnce(GetBilling, {
      billing: {
        on_trial: false,
        plan: 'free',
        subscribed: false,
      },
    })
    const { getByRole, getAllByRole, getByLabelText } = render(CustomDomain)

    await waitFor(() => {
      expect(getByLabelText(/enter your domain name/i)).toBeVisible()
    })

    const domainInput = getByLabelText(/enter your domain name/i)

    await waitFor(() => {
      expect(domainInput).toHaveValue('www.joesblog.storipress.dev')
    })

    const activateButton = getByRole('button', {
      name: /verify DNS Settings/i,
    })

    expect(activateButton).not.toBeDisabled()
    expect(getAllByRole('listitem')).toMatchSnapshot()
  })

  it('invalid domain', async () => {
    const { getByLabelText, getByText } = render(CustomDomain)

    await waitFor(() => {
      expect(getByLabelText(/enter your domain name/i)).toBeVisible()
    })

    const domainInput = getByLabelText(/enter your domain name/i)

    await waitFor(() => {
      expect(domainInput).toHaveValue('www.joesblog.storipress.dev')
    })

    await fireEvent.update(domainInput, 'https://www.joesblog.storipress.dev')
    await waitFor(() => {
      expect(getByText('Please remove the https:// from your domain')).toBeVisible()
    })
  })

  it('invalid submit', async () => {
    const { getByLabelText, queryByText } = render(CustomDomain)

    await waitFor(() => {
      expect(getByLabelText(/enter your domain name/i)).toBeVisible()
    })

    const domainInput = getByLabelText(/enter your domain name/i)

    await waitFor(() => {
      expect(domainInput).toHaveValue('www.joesblog.storipress.dev')
    })

    await fireEvent.update(domainInput, 'https://joesblog.storipress.dev')
    // expect(await findByText('Please remove the https:// from your domain')).toBeVisible()
    expect(queryByText('Domain activation requested')).not.toBeInTheDocument()
  })

  it('has get custom domain configuration but not yet activate', async () => {
    mockResponseOnce(GetSite, {
      site: {
        plan: 'free',
      },
    })
    mockResponseOnce(GetBilling, {
      billing: {
        on_trial: false,
        plan: 'free',
        subscribed: false,
      },
    })

    const { getByRole, getByLabelText } = render(CustomDomain)

    await waitFor(() => {
      expect(getByLabelText(/enter your domain name/i)).toBeVisible()
    })

    const domainInput = getByLabelText(/enter your domain name/i)

    await waitFor(() => {
      expect(domainInput).toHaveValue('www.joesblog.storipress.dev')
    })

    const activateButton = getByRole('button', {
      name: /verify DNS Settings/i,
    })
    expect(activateButton).not.toBeDisabled()
  })
})
