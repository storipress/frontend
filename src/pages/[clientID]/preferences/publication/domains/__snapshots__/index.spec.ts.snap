// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`setup publication domain > render custom domain and email 1`] = `
[
  <li>
    Record Type
  </li>,
  <li>
    Host
  </li>,
  <li>
    Value
  </li>,
  <li>
    TTL
  </li>,
  <li
    class="flex items-center gap-2"
  >
    <!--v-if-->
    <span>
      CNAME
    </span>
  </li>,
  <li
    class="w-full"
    data-testid="dns-host"
  >
    <div
      data-stub="hover-hint"
      offset="14"
      placement="right"
      popup-class="before:content-[''] before:absolute before:border before:-left-3 before:border-[6px] before:border-r-stone-700 before:border-y-transparent before:border-l-transparent delay-75"
      reference-class="w-full truncate"
    >
      <div
        data-stub="hover-hint-reference"
      >
        <span
          class="cursor-pointer px-1.5 py-1 duration-[50ms] hover:rounded hover:bg-gray-100"
        >
          www.joesblog
        </span>
      </div>
      <div
        data-stub="hover-hint-content"
      >
        Click to copy
      </div>
    </div>
  </li>,
  <li
    class="w-full"
    data-testid="dns-value"
  >
    <div
      data-stub="hover-hint"
      offset="14"
      placement="right"
      popup-class="before:content-[''] before:absolute before:border before:-left-3 before:border-[6px] before:border-r-stone-700 before:border-y-transparent before:border-l-transparent delay-75"
      reference-class="w-full truncate"
    >
      <div
        data-stub="hover-hint-reference"
      >
        <span
          class="cursor-pointer px-1.5 py-1 duration-[50ms] hover:rounded hover:bg-gray-100"
        >
          cdn.storipress.com
        </span>
      </div>
      <div
        data-stub="hover-hint-content"
      >
        Click to copy
      </div>
    </div>
  </li>,
  <li>
    10 Minutes
  </li>,
  <li
    class="flex items-center gap-2"
  >
    <!--v-if-->
    <span>
      CNAME
    </span>
  </li>,
  <li
    class="w-full"
    data-testid="dns-host"
  >
    <div
      data-stub="hover-hint"
      offset="14"
      placement="right"
      popup-class="before:content-[''] before:absolute before:border before:-left-3 before:border-[6px] before:border-r-stone-700 before:border-y-transparent before:border-l-transparent delay-75"
      reference-class="w-full truncate"
    >
      <div
        data-stub="hover-hint-reference"
      >
        <span
          class="cursor-pointer px-1.5 py-1 duration-[50ms] hover:rounded hover:bg-gray-100"
        >
          www.www.joesblog
        </span>
      </div>
      <div
        data-stub="hover-hint-content"
      >
        Click to copy
      </div>
    </div>
  </li>,
  <li
    class="w-full"
    data-testid="dns-value"
  >
    <div
      data-stub="hover-hint"
      offset="14"
      placement="right"
      popup-class="before:content-[''] before:absolute before:border before:-left-3 before:border-[6px] before:border-r-stone-700 before:border-y-transparent before:border-l-transparent delay-75"
      reference-class="w-full truncate"
    >
      <div
        data-stub="hover-hint-reference"
      >
        <span
          class="cursor-pointer px-1.5 py-1 duration-[50ms] hover:rounded hover:bg-gray-100"
        >
          cdn.storipress.com
        </span>
      </div>
      <div
        data-stub="hover-hint-content"
      >
        Click to copy
      </div>
    </div>
  </li>,
  <li>
    10 Minutes
  </li>,
  <li
    class="flex items-center gap-2"
  >
    <div
      class="inline"
      data-v-69d6dbbe=""
    >
      <span
        aria-label="show-tooltip"
        class="icon-question-mark-inverse trigger cursor-pointer text-stone-400"
        data-v-206d3c28=""
        data-v-69d6dbbe=""
      />
      <div
        class="tooltip layer-2 absolute rounded-md border border-stone-100 bg-white invisible"
        data-popper-escaped=""
        data-popper-placement="right-start"
        data-popper-reference-hidden=""
        data-v-69d6dbbe=""
        style="position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 0px);"
      >
        
        <div
          class="w-[21rem] p-4"
        >
          <div
            class="text-button mb-1.5 text-stone-800"
          >
            Email records
          </div>
          <div
            class="text-body text-stone-400"
          >
             These are used to send newsletters using your custom domain from Storipress. 
          </div>
        </div>
        
      </div>
    </div>
    <span>
      TXT
    </span>
  </li>,
  <li
    class="w-full"
    data-testid="dns-host"
  >
    <div
      data-stub="hover-hint"
      offset="14"
      placement="right"
      popup-class="before:content-[''] before:absolute before:border before:-left-3 before:border-[6px] before:border-r-stone-700 before:border-y-transparent before:border-l-transparent delay-75"
      reference-class="w-full truncate"
    >
      <div
        data-stub="hover-hint-reference"
      >
        <span
          class="cursor-pointer px-1.5 py-1 duration-[50ms] hover:rounded hover:bg-gray-100"
        >
          20230602053200pm._domainkey.www.joesblog
        </span>
      </div>
      <div
        data-stub="hover-hint-content"
      >
        Click to copy
      </div>
    </div>
  </li>,
  <li
    class="w-full"
    data-testid="dns-value"
  >
    <div
      data-stub="hover-hint"
      offset="14"
      placement="right"
      popup-class="before:content-[''] before:absolute before:border before:-left-3 before:border-[6px] before:border-r-stone-700 before:border-y-transparent before:border-l-transparent delay-75"
      reference-class="w-full truncate"
    >
      <div
        data-stub="hover-hint-reference"
      >
        <span
          class="cursor-pointer px-1.5 py-1 duration-[50ms] hover:rounded hover:bg-gray-100"
        >
          k=rsa; p=MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQC1RG+F584xu1j9xHYUrLCCtM+eb/yF1gww8oB11fZzVjIIcF0hgdyaCai2E6RlVALphG7dtu0qh/Jf6rA8oFAUZDuBGurzohkMzwrKot81RUd6y30h9fTYICeGBZnBMsmMCc5ZNL2wIsjfau5WOUETtC7YNz0OwLvTTR1Wu2s0swIDAQAB
        </span>
      </div>
      <div
        data-stub="hover-hint-content"
      >
        Click to copy
      </div>
    </div>
  </li>,
  <li>
    10 Minutes
  </li>,
  <li
    class="flex items-center gap-2"
  >
    <div
      class="inline"
      data-v-69d6dbbe=""
    >
      <span
        aria-label="show-tooltip"
        class="icon-question-mark-inverse trigger cursor-pointer text-stone-400"
        data-v-206d3c28=""
        data-v-69d6dbbe=""
      />
      <div
        class="tooltip layer-2 absolute rounded-md border border-stone-100 bg-white invisible"
        data-popper-escaped=""
        data-popper-placement="right-start"
        data-popper-reference-hidden=""
        data-v-69d6dbbe=""
        style="position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 0px);"
      >
        
        <div
          class="w-[21rem] p-4"
        >
          <div
            class="text-button mb-1.5 text-stone-800"
          >
            Email records
          </div>
          <div
            class="text-body text-stone-400"
          >
             These are used to send newsletters using your custom domain from Storipress. 
          </div>
        </div>
        
      </div>
    </div>
    <span>
      CNAME
    </span>
  </li>,
  <li
    class="w-full"
    data-testid="dns-host"
  >
    <div
      data-stub="hover-hint"
      offset="14"
      placement="right"
      popup-class="before:content-[''] before:absolute before:border before:-left-3 before:border-[6px] before:border-r-stone-700 before:border-y-transparent before:border-l-transparent delay-75"
      reference-class="w-full truncate"
    >
      <div
        data-stub="hover-hint-reference"
      >
        <span
          class="cursor-pointer px-1.5 py-1 duration-[50ms] hover:rounded hover:bg-gray-100"
        >
          20230602053200pm._domainkey.www.joesblog
        </span>
      </div>
      <div
        data-stub="hover-hint-content"
      >
        Click to copy
      </div>
    </div>
  </li>,
  <li
    class="w-full"
    data-testid="dns-value"
  >
    <div
      data-stub="hover-hint"
      offset="14"
      placement="right"
      popup-class="before:content-[''] before:absolute before:border before:-left-3 before:border-[6px] before:border-r-stone-700 before:border-y-transparent before:border-l-transparent delay-75"
      reference-class="w-full truncate"
    >
      <div
        data-stub="hover-hint-reference"
      >
        <span
          class="cursor-pointer px-1.5 py-1 duration-[50ms] hover:rounded hover:bg-gray-100"
        >
          pm.mtasv.net
        </span>
      </div>
      <div
        data-stub="hover-hint-content"
      >
        Click to copy
      </div>
    </div>
  </li>,
  <li>
    10 Minutes
  </li>,
]
`;
